pynicotine/slskproto.py

```diff
@@ -311,44 +317,159 @@ def _bind_listen_port(self):
+            ("adapter_name", wintypes.LPSTR),
+            ("first_unicast_address", POINTER(IP_ADAPTER_UNICAST_ADDRESS)),
+            ("first_anycast_address", wintypes.LPVOID),
+            ("first_multicast_address", wintypes.LPVOID),
+            ("first_dns_server_address", wintypes.LPVOID),
+            ("dns_suffix", wintypes.LPWSTR),
+            ("description", wintypes.LPWSTR),
+            ("friendly_name", wintypes.LPWSTR)
+        ]
+
+        interface_addresses = {}
+        address_buffer_size = wintypes.ULONG(1024 * 15)
+        return_value = ERROR_BUFFER_OVERFLOW
+
+        while return_value == ERROR_BUFFER_OVERFLOW:
+            address_buffer = create_string_buffer(address_buffer_size.value)
+            return_value = windll.Iphlpapi.GetAdaptersAddresses(
+                AF_INET,
+                (GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_DNS_SERVER),
+                None,
+                byref(address_buffer),
+                byref(address_buffer_size),
+            )
+
+        if return_value:
+            log.add_debug("Failed to get list of network interfaces. Error code: %s", return_value)
+            return interface_addresses
+
+        adapter_addresses = IP_ADAPTER_ADDRESSES.from_buffer(address_buffer)
+
+        while True:
+            if adapter_addresses.first_unicast_address:
+                interface_name = adapter_addresses.friendly_name
+                socket_address = adapter_addresses.first_unicast_address[0].address
+                interface_addresses[interface_name] = socket.inet_ntoa(socket_address.lp_sockaddr[0].sin_addr)
+
+            if not adapter_addresses.next:
+                break
+
+            adapter_addresses = adapter_addresses.next[0]
+
+        return interface_addresses
+
+    @staticmethod
+    def _get_network_interface_addresses_other():
+        """ Returns a dictionary of network interface names and IP addresses (other OSes) """
+
+        interface_addresses = {}
 
         try:
-            import fcntl
-            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            interface_name_index = socket.if_nameindex()
 
-            ip_if = fcntl.ioctl(sock.fileno(),
-                                SIOCGIFADDR,
-                                struct.pack("256s", if_name.encode()[:15]))
+        except (AttributeError, OSError) as error:
+            log.add_debug("Failed to get list of network interfaces. Error: %s", error)
+            return interface_addresses
 
-            ip_address = socket.inet_ntoa(ip_if[20:24])
+        for _i, interface_name in interface_name_index:
+            try:
+                import fcntl  # pylint: disable=import-error
 
-        except ImportError:
-            ip_address = None
+                with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
+                    siocgifaddr = 0x8915 if sys.platform == "linux" else 0xc0206921  # 0xc0206921 for *BSD, macOS
```

Comment:
Why is not constant anymore?

Also, fnctl only needs to be imported once.

Reply:
> Also, fnctl only needs to be imported once.

I can't import fnctl globally, it's not available on Windows.